File: 0766-toeplitz-matrix/0766-toeplitz-matrix.js
/**
 * @param {number[][]} matrix
 * @return {boolean}
 */
const isToeplitzMatrix = function(matrix) {
    const m = matrix.length, n = matrix[0].length;
    
    const isIB = (r, c) => r >= 0 && c >= 0 && r < m && c < n;
    
    const checkDiagonal = (r, c) => {
        const val = matrix[r][c];
        while (isIB(r + 1, c + 1)) {
            if (matrix[++r][++c] !== val) return false;
        }
        return true;
    }
    
    for (let r = 0; r < m; r++) {
        if (!checkDiagonal(r, 0)) return false;
    }
    
    for (let c = 1; c < n; c++) {
        if (!checkDiagonal(0, c)) return false;
    }
    
    return true;
};

File: 0766-toeplitz-matrix/0766-toeplitz-matrix.cs
public class Solution {
    public bool IsToeplitzMatrix(int[][] matrix) {
        int m = matrix.Length, n = matrix[0].Length;
        
        bool isInbounds(int r, int c) {
            return r >= 0 && c >= 0 && r < m && c < n;
        }
        
        for (int r = 0; r < m; r++)
        {
            int rr = r, cc = 0, val = matrix[r][0];
            
            while (isInbounds(rr + 1, cc + 1))
            {
                if (matrix[++rr][++cc] != val) return false;
            }
        }
        
        for (int c = 0; c < n; c++)
        {
            int rr = 0, cc = c, val = matrix[0][c];
            
            while (isInbounds(rr + 1, cc + 1))
            {
                if (matrix[++rr][++cc] != val) return false;
            }
        }
        
        return true;
    }
}